### PR TITLE
(CDAP-16308) Introduce a local disk cache for Location

### PR DESCRIPTION
**Pull request by chtyim**
- Cache remote Location to local disk
- Improve preview and app deployment performance

### CODE REVIEW

**Comment by chtyim** [comment]
Build https://builds.cask.co/browse/CDAP-DUT7109

File: cdap-common/src/main/java/io/cdap/cdap/common/guice/DFSLocationModule.java [hunk 1]
```diff
@@ -66,16 +73,28 @@ private LocationFactoryProvider(CConfiguration cConf, Configuration hConf,
...
+        lf = new FileContextLocationFactory(hConf, namespace);
+      } else {
+        // In non hadoop secure mode, use the static file context, which operates as single user.
+        lf = new InsecureFileContextLocationFactory(hConf, namespace, staticFileContextProvider.get());
       }
 
-      // In non hadoop secure mode, use the static file context, which operates as single user.
-      return new InsecureFileContextLocationFactory(hConf, namespace, staticFileContextProvider.get());
+      String locationCachePath = cConf.get(Constants.LOCATION_CACHE_PATH);
+      Path cachePath = Strings.isNullOrEmpty(locationCachePath) ? null : Paths.get(locationCachePath).toAbsolutePath();
```
Comment (albertshau) [inline cdap-common/src/main/java/io/cdap/cdap/common/guice/DFSLocationModule.java]: would be a little simpler to return lf if locationCachePath is null before this line

File: cdap-common/src/main/java/io/cdap/cdap/common/io/CachingLocation.java [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+      // If fail to open stream from the cache, try to regenerate it.
+      Files.deleteIfExists(cachePath);
+    }
+
+    try (InputStream input = delegate.getInputStream()) {
+      Files.createDirectories(cachePath.getParent());
+      Path tmpPath = Files.createTempFile(cachePath.getParent(), cachePath.getFileName().toString(), ".tmp",
+                                          PosixFilePermissions.asFileAttribute(
+                                            PosixFilePermissions.fromString("rw-------")));
+      Files.copy(input, tmpPath, StandardCopyOption.REPLACE_EXISTING);
```
Comment (albertshau) [inline cdap-common/src/main/java/io/cdap/cdap/common/io/CachingLocation.java:119]: do we need to delete the file at the tmpPath? Or is that somehow handled by Files.createTempFile?

Comment (chtyim) [reply cdap-common/src/main/java/io/cdap/cdap/common/io/CachingLocation.java:119]: It is handled by `StandardCopyOption.REPLACE_EXISTING`.

**Review by albertshau — APPROVED** [review]

**Pull request merged**